### PR TITLE
Users/torivara/virtual network gateways update

### PR DESCRIPTION
**Pull request by torivara**
# Change

Virtual Network Gateway with ExpressRoute type will not work if Public IP is set as zonal and gateway is set as regional. This means that the default settings will fail in some cases because PIP is zonal and gateway is not. Only tested with Express Route Gateway.

The proposed change is an update to the default value of publicIpZones and a conditional setting of zones on the automatically created pip.

Maybe this module should also support already existing pip-resources?

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: arm/Microsoft.Network/virtualNetworkGateways/deploy.bicep [hunk 1]
```diff
@@ -290,7 +288,7 @@ resource virtualGatewayPublicIP 'Microsoft.Network/publicIPAddresses@2021-02-01'
     publicIPPrefix: ((!empty(publicIPPrefixId)) ? publicIPPrefix : json('null'))
     dnsSettings: ((length(virtualGatewayPipName_var) == length(domainNameLabel)) ? json('{"domainNameLabel": "${domainNameLabel[index]}"}') : json('null'))
   }
-  zones: publicIpZones
+  zones: contains(zoneRedundantSkus, virtualNetworkGatewaySku) ? publicIpZones : json('null')
```
Comment (MrMCake) [inline arm/Microsoft.Network/virtualNetworkGateways/deploy.bicep:291]: Hey @torivara just one comment, please change `json('null')` to just `null`.

Comment (MrMCake) [reply arm/Microsoft.Network/virtualNetworkGateways/deploy.bicep:291]: Bicep magic ;)

Comment (torivara) [reply arm/Microsoft.Network/virtualNetworkGateways/deploy.bicep:291]: I was just following the rest of the document, @MrMCake. Should all json('null') be changed to just null? There are 19 occurences of json('null') the vnetgateway bicep file alone.

Comment (MrMCake) [reply arm/Microsoft.Network/virtualNetworkGateways/deploy.bicep:291]: @torivara  If so, please do, yes. The more consistent the code is the better. Will check the repo in general today to make sure this happens in all other places too.
We're saving ink ;)

Comment (MrMCake) [reply arm/Microsoft.Network/virtualNetworkGateways/deploy.bicep:291]: 606 results in 87 files. Nice. Will do that after we merged your PR

Comment (torivara) [reply arm/Microsoft.Network/virtualNetworkGateways/deploy.bicep]: Search/Replaced all json('null') with null now. This looks like it is working. Tested deployment to my own sub, and there are no validation errors. Thank you for the suggestion!

File: arm/Microsoft.Network/virtualNetworkGateways/readme.md [hunk 2]
```diff
@@ -33,7 +33,7 @@ This module deploys a Virtual Network Gateway.
 | `metricsToEnable` | array | `[AllMetrics]` | `[AllMetrics]` | Optional. The name of metrics that will be streamed. |
 | `publicIpLogsToEnable` | array | `[DDoSProtectionNotifications, DDoSMitigationFlowLogs, DDoSMitigationReports]` | `[DDoSProtectionNotifications, DDoSMitigationFlowLogs, DDoSMitigationReports]` | Optional. The name of logs that will be streamed. |
 | `publicIPPrefixId` | string |  |  | Optional. Resource Id of the Public IP Prefix object. This is only needed if you want your Public IPs created in a PIP Prefix. |
-| `publicIpZones` | array | `[1]` |  | Optional. Specifies the zones of the Public IP address. |
+| `publicIpZones` | array | `[]` |  | Optional. Specifies the zones of the Public IP address. Basic IP SKU does not support Availability Zones.|
```
Comment (MariusStorhaug) [inline arm/Microsoft.Network/virtualNetworkGateways/readme.md:36]: If you are updating the parameter description, please also do so in the `deploy.bicep` file. In the contribution guide we also have info about a script under utilities/tools folder to generate the readme from the bicep file.

**Comment by MrMCake** [comment]
Resolved a conflcit for you

File: arm/Microsoft.Network/virtualNetworkGateways/readme.md (hunk 2 above)
Comment (torivara) [reply arm/Microsoft.Network/virtualNetworkGateways/readme.md:36]: Thank you for pointing that out, @MariusStorhaug! Should be fixed now. I have not seen any information on the tools available. Where can I find this in the contribution guide?

Comment (MariusStorhaug) [reply arm/Microsoft.Network/virtualNetworkGateways/readme.md]: Sorry, its in the wiki, but not the contribution guide.
https://github.com/Azure/ResourceModules/wiki/ModulesDesign#structure

Comment (MariusStorhaug) [reply arm/Microsoft.Network/virtualNetworkGateways/readme.md]: There is a reference here to a script. Though the link to it is wrong. Its under: utilities\tools\Set-ModuleReadMe.ps1

**Review by MariusStorhaug — APPROVED** [review]
All changes in + the merge from main, so all in good shape this time around. Approved.

**Pull request merged**